indico/web/client/js/utils/date.js

```diff
@@ -160,3 +160,71 @@ export function serializeDateTimeRange(startDt, endDt) {
     ? `${startDate} ${startTime} - ${endTime}`
     : `${startDate} ${startTime} - ${endDate} ${endTime}`;
 }
+
+export function getToday() {
+  const now = new Date();
+  now.setHours(0, 0, 0, 0);
+  return now;
+}
+
+export function toDateString(date) {
+  // We do not use ISO date format because it will be interpreted as UTC
+  return date?.toDateString() || '';
+}
+
+export function toOptionalDate(dateString) {
+  const date = new Date(dateString);
+  if (!date.getTime()) {
+    return;
+  }
+  date.setHours(0, 0, 0, 0);
+  return date;
+}
+
+export function isSameDate(x, y) {
+  return x?.toDateString() === y?.toDateString();
+}
+
+class DateRangeBase {
+  constructor(start, end) {
+    this.start = toOptionalDate(start);
+    this.end = toOptionalDate(end);
+  }
+
+  get isInvalid() {
+    return !this.start && !this.end;
+  }
+
+  startsWith(date) {
+    return isSameDate(date, this.start);
+  }
+
+  endsWith(date) {
+    return isSameDate(date, this.end);
+  }
+}
+
+export class DateRange extends DateRangeBase {
+  includes(date) {
+    if (this.isInvalid) {
+      return false;
+    }
+    if (!this.start) {
+      return isSameDate(date, this.end);
+    }
+    if (!this.end) {
+      return isSameDate(date, this.start);
+    }
+    return date >= this.start && date <= this.end;
+  }
+}
+
+export class OpenDateRange extends DateRangeBase {
+  includes(date) {
+    if (this.isInvalid) {
+      return true;
+    }
+
+    return date >= (this.start || -Infinity) && date <= (this.end || Infinity);
+  }
+}
```

Comment:
Is there a reason you decided not to use moment for these? Most, if not all of the date-related functions are already available so moment is what we typically use in most places

Reply:
I could also ask the same about *using* moment. Most date-related functionality is already available in the browser.

Reply:
I think one risk with rewriting things from scratch is inconsistency, e.g. with things like displaying localized month/weekday names (getting them from the moment locale data vs getting them from Intl).

Also, I agree so far that if we were to start developing Indico now, we almost certainly would not be using moment - after all, even its developers say that it's kind of obsolete nowadays because native support is good enough.

On the other hand, there are so many places in Indico already using moment (because in the past, core JS/browser utils lacked lots of functionality) that it'll probably stay for a looooong time. After all, it's not a bad library nor one that encourages people to write hard-to-maintain code (unlike e.g. jQuery).

So in the end it's more consistent to use moment for such things, and just use Intl in places where moment does not have the right functionality.

Reply:
Tbh, I see no risk of such inconsistency here, because:

1. The entire custom element doesn't rely on moment at all
2. These abstractions deal with date comparisons. If there's any inconsistency, then there's a bug in moment.
3. The function `toDateString()` is *not* concerned with user-facing formatting. It is used for serializing dates into strings with the added fallback to empty string.

Lastly, Moment doesn't appear to provide an equivalent functionality without using a 3rd party add-on library.